### PR TITLE
57 handle a lot of pipes

### LINKED ISSUE #57: Handle a lot of pipes
## Description
need to handle a lot of pipes (about 1000 times?)

### current implementation

```bash
/bin/echo Hello | /bin/cat ...(126 times or more) |  /bin/cat
minishell: pipe: Too many open files
```

### On bash5.2

```bash
/bin/echo Hello | /bin/cat ...(126 times or more) |  /bin/cat
Hello
```

around 1088 times
```bash
bash: fork: retry: Resource temporarily unavailable
bash: child setpgid (42146 to 41059): Operation not permitted
Hello
```

### Generate experiment commands

Create and Edit `cmd.rb`

```ruby
cmd = "/bin/echo Hello"
pipe_cmd = " | /bin/cat"
iters = 127

iters.times do
  cmd << pipe_cmd
end

puts cmd
```

Execute ruby script

```bash
ruby cmd.rb | pbcopy
```

### PR DESCRIPTION
**Pull request by rask24**
fix #57

- **Refactor parser tests**
- **Update parser tests for new pipelie structure**
- **Fix parse_pipeline**

## Summary by CodeRabbit

- **New Features**
  - Introduced a centralized execution function for different types of AST nodes, improving command execution logic.
  - Added a new function to manage pipeline node construction in the AST.

- **Bug Fixes**
  - Enhanced error handling in the command execution process and during redirect management.

- **Tests**
  - Added new end-to-end tests for handling multiple commands in a pipeline.
  - Refactored unit tests for better readability and maintainability by standardizing token list construction.

- **Chores**
  - Updated the Makefile to reflect changes in source file references.
  - Removed outdated command not found macro from the utilities header.

### CODE REVIEW

**Comment by rask24** [comment]
@alex216 Now, minishell can handle 1000 pipes. But pipe fd management is not yet complete. I'll address this in #90.

File: src/exec/execute_pipeline.c [hunk 1]
```diff
@@ -18,25 +18,52 @@
...
+		t_pipeline_conf *conf, int *pipe_fd)
+{
+	if (conf->fd_in != STDIN_FILENO)
+		close(conf->fd_in);
+	conf->fd_in = pipe_fd[PIPE_READ];
+	if (node->type != AST_PIPE)
+		conf->fd_out = STDOUT_FILENO;
+	return (execute_ast_node(node, env_list, conf));
+}
+
```
[the remark this answers is not in the document]
Comment (rask24) [reply src/exec/execute_pipeline.c:45]: Exactly. I'll fix it.

File: sample [hunk 2]
```diff
@@ -0,0 +1,10 @@
+- ** Fix exexute_pipline
```
[the remark this answers is not in the document]
Comment (rask24) [reply sample:1]: I'll delete this file

File: src/exec/execute_pipeline.c [hunk 3]
```diff
@@ -18,25 +18,56 @@
...
+{
+	int	ret;
+
+	if (conf->fd_in != STDIN_FILENO)
+		close(conf->fd_in);
+	conf->fd_in = pipe_fd[PIPE_READ];
+	if (node->type != AST_PIPE)
+		conf->fd_out = STDOUT_FILENO;
+	ret = execute_ast_node(node, env_list, conf);
+	close(pipe_fd[PIPE_READ]);
```
Comment (alex216) [inline src/exec/execute_pipeline.c:46]: redundant `close()`, since Line40,41 has done it already.
```suggestion
```

**Review by alex216 — COMMENTED** [review]
@rask24 please check the comment!

File: src/exec/execute_pipeline.c (hunk 3 above)
Comment (rask24) [reply src/exec/execute_pipeline.c:46]: It is necessary.
Usually `pipe_fd[PIPE_READ]` will be closed at line 41
but `pipe_fd[PIPE_READ]` of last command in the pipeline won't be closed at the line.
So line 46 is necessary.
@alex216

Comment (rask24) [reply src/exec/execute_pipeline.c:46]: ## Experiment

### Current implementation

``` bash
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello

# In another terminal

❯ lsof -c minishell
COMMAND     PID       USER   FD   TYPE DEVICE SIZE/OFF     NODE NAME
minishell 59046 reiyaasuke  cwd    DIR   1,14      800 68945239 /Users/reiyaasuke/clang/minishell
minishell 59046 reiyaasuke  txt    REG   1,14   383704 75835580 /Users/reiyaasuke/clang/minishell/minishell
minishell 59046 reiyaasuke    0u   CHR  16,37 0t160534     6961 /dev/ttys037
minishell 59046 reiyaasuke    1u   CHR  16,37 0t160534     6961 /dev/ttys037
minishell 59046 reiyaasuke    2u   CHR  16,37 0t160534     6961 /dev/ttys037
```

There are no fd leaks.

### Proposed implementation (delete line 46)

``` bash
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello

# In another terminal
❯ lsof -c minishell
COMMAND     PID       USER   FD   TYPE             DEVICE SIZE/OFF     NODE NAME
minishell 59572 reiyaasuke  cwd    DIR               1,14      800 68945239 /Users/reiyaasuke/clang/minishell
minishell 59572 reiyaasuke  txt    REG               1,14   383704 75837356 /Users/reiyaasuke/clang/minishell/minishell
minishell 59572 reiyaasuke    0u   CHR              16,37 0t171299     6961 /dev/ttys037
minishell 59572 reiyaasuke    1u   CHR              16,37 0t171299     6961 /dev/ttys037
minishell 59572 reiyaasuke    2u   CHR              16,37 0t171299     6961 /dev/ttys037
minishell 59572 reiyaasuke    3   PIPE 0xa093add1cdd27c2d    16384
minishell 59572 reiyaasuke    4   PIPE 0x7d67927a3a8532cf    16384
minishell 59572 reiyaasuke    5   PIPE 0x611672c7b0505e43    16384
minishell 59572 reiyaasuke    6   PIPE 0xe2ade9d1fd4af972    16384
```

There are some fd leaks.

**Review by alex216 — APPROVED** [review]
@rask24 ok!

**Pull request merged**